### ISSUE TITLE
Add  labels to navigable components

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/molecules-and-light#163]

**Comment by terracoda** [comment]
@jessegreenberg, I might already have some labels for this sim here:
https://docs.google.com/document/d/1r9D3vq4QvmmXkQ5G6HexpybUv_vL6LrYFb3wgPfUdWg/edit

**Comment by jessegreenberg** [comment]
For radio buttons, use exac labels that are there.

For slider, use [light name] + Light source.
Add play, pause, step labels to  those buttons.
Add popup ARIA attribute to "Show light spectrum" button, and include text as accessible label.

**Comment by jessegreenberg** [comment]
@zepumph would you mind working on this? Sorry about the chicken scratch in https://github.com/phetsims/molecules-and-light/issues/163#issuecomment-347611563. The design doc @terracoda added is much nicer. Let me know if you have any questions.

**Comment by terracoda** [comment]
@jessegreenberg, I think I was wrong about the pop-up part for the Show light spectrum, button.

**Comment by terracoda** [comment]
The Show light spectrum button would be more similar to the Keyboard Shortcuts button than the PhET Menu button.

Is there a general "Has pop-up" that would be beneficial for both cases?

**Comment by jessegreenberg** [comment]
@terracoda would this be used?
https://www.w3.org/TR/wai-aria-1.1/#aria-haspopup

**Comment by terracoda** [comment]
Yes @jessegreenberg , that is the proper attribute that we could use for any content that gets popped up in a dialog.

The PhET Menu button, I think is different because it pops up an actual menu, but it seems like it may also be in this category.

**Comment by terracoda** [comment]
What is the difference in code between the PhET Menu button and the Keyboard Shortcuts button? They do sound different as currently implemented.

**Comment by jessegreenberg** [comment]
The PhET menu button is already using this attribute. https://www.w3.org/TR/wai-aria-1.1/#aria-haspopup includes menus in their recommendation. @terracoda should we add this to the keyboard help button and the "Show Light Spectrum" button in this sim?

**Comment by terracoda** [comment]
@jessegreenberg, I think if browser and AT support is good, it would be good to do the same for dialogs that pop-up.

It gives a user the awareness that they are in separate space than the sim.

**Comment by jessegreenberg** [comment]
Sounds good! It was tested when we deployed john-travoltage so I am confident in it. If it is too verbose we can easily remove it later.

**Comment by terracoda** [comment]
Sounds good.

**Comment by zepumph** [comment]
The a11y view on master should be correct. Please review @jessegreenberg

**Comment by phet-steele** [comment]
@zepumph did your commits cause this in a handful of repos?:

```
Error: Assertion failed: Undefined not allowed for Node key: accessibleLabel
    at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/assert/js/assert.js:22:13)
    at https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/scenery/js/nodes/Node.js?bust=1511946659414:4795:19
    at u (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/sherpa/lib/lodash-4.17.4.min.js?bust=1511946659414:5:432)
    at Function.ru (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/sherpa/lib/lodash-4.17.4.min.js?bust=1511946659414:67:656)
    at StepForwardButton.mutate (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/scenery/js/nodes/Node.js?bust=1511946659414:4793:9)
    at StepForwardButton.RoundButtonView [as constructor] (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/sun/js/buttons/RoundButtonView.js?bust=1511946659414:135:10)
    at StepForwardButton.RoundPushButton [as constructor] (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/sun/js/buttons/RoundPushButton.js?bust=1511946659414:49:21)
    at StepForwardButton.StepButton [as constructor] (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/scenery-phet/js/buttons/StepButton.js?bust=1511946659414:81:21)
    at new StepForwardButton (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/scenery-phet/js/buttons/StepForwardButton.js?bust=1511946659414:24:16)
    at new SOMPlayPauseStepControl (https://bayes.colorado.edu/continuous-testing/snapshot-1511934071545/states-of-matter/js/common/view/SOMPlayPauseStepControl.js?bust=1511946659414:41:22)
```

**Comment by zepumph** [comment]
Yes the above is because of me. In SceneryPhet I update a few of the a11y strings to be object with a "value" named key that holds the string. This breaks xss fuzzing. It will be fixed immediately in https://github.com/phetsims/a11y-research/issues/65.

**Comment by jessegreenberg** [comment]
@zepumph this is looking great! I had a few questions:

- [ ] Could you please add a label for the close button in the light spectrum dialog?
- [ ] Sorry, it was requested later in https://github.com/phetsims/molecules-and-light/issues/163#issuecomment-347615642, but could you please add the `aria-haspopup` attribute to  the "Show Light Spectrum" button?
- [ ] In the interest of making i18n easier farther down the road, could you please create a string pattern  and use `StringUtils.fillIn` for this concat?
```js
this.emissionRateControlSlider.accessibleLabel = lightSourceName + ' ' + lightSourceString;
```
- [ ] Could you please move all new labels into the new MoleculesAndLightA11yStrings.js?
- [ ] Thanks for the many formatting fixes! I happened to catch one removal of spacing, is this correct?
```js
-      rotation: ( options.direction === 'forward' ) ? 0 : Math.PI
+      rotation: (options.direction === 'forward') ? 0 : Math.PI
```
- [ ] In https://github.com/phetsims/sun/commit/f408f8bec052bb90631e8c12c45746abd96fb70c, could these options be defined as defaults in opts? Why do we need to modify opts?
```js
// a11y create the label for the radio button
 if ( currentContent.accessibleLabel ) {
   opts.accessibleLabel = currentContent.accessibleLabel;
   opts.labelTagName = 'label';
   opts.parentContainerTagName = 'div';
  }
```

**Comment by jessegreenberg** [comment]
> but could you please add the aria-haspopup attribute to the "Show Light Spectrum" button?

Sorry! You already did it!

**Comment by zepumph** [comment]
>Thanks for the many formatting fixes! I happened to catch one removal of spacing, is this correct?

Nope, but I'll fix it. See https://github.com/phetsims/phet-info/issues/62

**Comment by zepumph** [comment]
>In phetsims/sun@f408f8b, could these options be defined as defaults in opts? Why do we need to modify opts?

This was the best way I could think of implementing so that we never got undefined problems. If you don't pass an accessibleLabel to a radio button group member when creating a RadioButtonGroup, but you always have that option add whatever the value of `currentContent.accessibleLabel`, then you end up passing Node `{ accessibleLabel: undefined }`, which will cause an error because node setters can't set a value to undefined.

**Comment by zepumph** [comment]
Everything else is fixed but the above.

**Comment by jessegreenberg** [comment]
Got it, thanks for clarifying. I was wondering about
```js
        // a11y
        accessibleLabel: currentContent.accessibleLabel || null,
        labelTagName: 'label',
        parentContainerTagName: 'div'
```

But it is really splitting hairs. I see there is some symmetry with phet-io on the preceding lines, and this way we don't hit
```js
        labelTagName: 'label',
        parentContainerTagName: 'div'
```
unless we need to.

I verified everything else is fixed, it is sounding sharp on NVDA. @zepumph good to close?

**Comment by zepumph** [comment]
Now I'm confused. The only usage of `accessibleLabel || null` that I see is in `VerticalAquaRadioButtonGroup.js`, but I didn't put that in.

```js
      var radioButton = new AquaRadioButton( items[ i ].property, items[ i ].value, content, _.extend( {}, options.radioButtonOptions, {
        radius: options.radius,
        tandem: items[ i ].tandemName ? options.tandem.createTandem( items[ i ].tandemName ) : Tandem.required,
        accessibleLabel: items[ i ].accessibleLabel || null
      } ) );
```

Now that we have done work to RadioButtonGroup, @jessegreenberg does this look appropriate as well? If so please close.

**Comment by jessegreenberg** [comment]
I believe I added that code. I am more OK with the usage in VerticalAquaRadioButtonGroup since it is the only option used, we could also have

```js
if ( items[ i ].accessibleLabel ) {
  radioButton.accessibleLabel = items[ i ].accessibleLabel;
)
```

But I think the way it is is OK. Ill go ahead and close, but @zepumph if you are concerned please feel free to reopen.
